File: examples/secure/python/path_traversal_fixed.py
"""
✅ SECURE — Path Traversal Fix (CWE-22)
https://cwe.mitre.org/data/definitions/22.html

Principle: Canonicalize the path and verify it stays inside the allowed
           directory before any file operation.

This file is the fixed version of:
    ../../vulnerable/python/path_traversal.py

What changed and why:
    1. Resolved the requested path to its absolute, canonical form with
       os.path.realpath() — this collapses "..", symlinks, and other
       tricks an attacker might use.
    2. Verified that the canonical path starts with the canonical upload
       directory (os.path.commonpath check).  If the resolved path escapes
       the allowed directory, the request is rejected.
    3. Rejected filenames that contain path separators or are empty.

Detected by: CodeQL (python/path-injection)
"""

import os
from flask import Flask, request, send_file, jsonify

app = Flask(__name__)

UPLOAD_DIR = os.path.realpath("/uploads")


def _safe_path(base_dir: str, user_input: str) -> str | None:
    """Return the canonical path if it is inside base_dir, else None."""
    if not user_input:
        return None
    # Resolve the full path and collapse any ".." or symlink tricks.
    requested = os.path.realpath(os.path.join(base_dir, user_input))
    # Verify the resolved path is still inside the allowed directory.
    if os.path.commonpath([base_dir, requested]) != base_dir:
        return None
    return requested


# GOOD: Path is canonicalized and confined to UPLOAD_DIR before serving.
@app.route("/download")
def download():
    filename = request.args.get("file", "")

    file_path = _safe_path(UPLOAD_DIR, filename)
    if file_path is None:
        return jsonify({"error": "Invalid file path"}), 400

    if not os.path.isfile(file_path):
        return jsonify({"error": "File not found"}), 404

    return send_file(file_path)


# GOOD: Path is canonicalized and confined before reading.
@app.route("/view")
def view_file():
    filename = request.args.get("file", "")

    file_path = _safe_path(UPLOAD_DIR, filename)
    if file_path is None:
        return jsonify({"error": "Invalid file path"}), 400

    if not os.path.isfile(file_path):
        return jsonify({"error": "File not found"}), 404

    with open(file_path, "r") as f:
        content = f.read()

    return jsonify({"filename": filename, "content": content})


# GOOD: Directory path is canonicalized and confined before listing.
@app.route("/list")
def list_files():
    directory = request.args.get("dir", "")

    dir_path = _safe_path(UPLOAD_DIR, directory) if directory else UPLOAD_DIR
    if dir_path is None:
        return jsonify({"error": "Invalid directory path"}), 400

    if not os.path.isdir(dir_path):
        return jsonify({"error": "Directory not found"}), 404

    files = os.listdir(dir_path)
    return jsonify({"directory": directory, "files": files})


if __name__ == "__main__":
    app.run(debug=True, port=5000)

File: examples/secure/csharp/PathTraversalFixed.cs
/**
 * ✅ SECURE — Fixed Path Traversal Example (C#)
 *
 * Security Principle : Input Validation + Canonical Path Verification
 * CWE Addressed      : CWE-22  (Improper Limitation of a Pathname to a Restricted Directory)
 * CodeQL query        : cs/path-injection
 *
 * What Changed:
 *   1. Strip path components with Path.GetFileName() — removes any
 *      directory traversal sequences (../, ..\ , absolute paths).
 *   2. Reject filenames containing ".." or path separators as an
 *      additional defence-in-depth check.
 *   3. Resolve the canonical (full) path and verify it starts with
 *      the allowed base directory, preventing symlink or encoding tricks.
 *
 * Compare with: examples/vulnerable/csharp/PathTraversal.cs
 */

using System;
using System.IO;
using Microsoft.AspNetCore.Mvc;

namespace GhasPack.Secure.CSharp
{
    [ApiController]
    [Route("api/[controller]")]
    public class PathTraversalFixedController : ControllerBase
    {
        private static readonly string UploadsDirectory =
            Path.GetFullPath("/uploads");

        // GET /api/pathtraversalfixed?file=report.pdf
        [HttpGet]
        public IActionResult DownloadFile([FromQuery] string file)
        {
            // FIX #1: Reject null or empty filenames
            if (string.IsNullOrWhiteSpace(file))
            {
                return BadRequest("Missing required parameter: file");
            }

            // FIX #2: Strip directory components — Path.GetFileName returns
            // only the filename portion, discarding any leading path segments.
            // "../../etc/passwd" → "passwd", "C:\Windows\win.ini" → "win.ini"
            string sanitizedName = Path.GetFileName(file);

            // FIX #3: Defence-in-depth — reject if the original input contained
            // traversal characters.  This catches edge cases and makes the
            // security intent explicit in the code.
            if (file.Contains("..") ||
                file.Contains(Path.DirectorySeparatorChar) ||
                file.Contains(Path.AltDirectorySeparatorChar))
            {
                return BadRequest("Invalid filename.");
            }

            // FIX #4: Resolve the full canonical path and verify it is still
            // inside the allowed directory.  This protects against symlinks,
            // URL-encoded path separators, or other bypass techniques.
            string fullPath = Path.GetFullPath(
                Path.Combine(UploadsDirectory, sanitizedName));

            if (!fullPath.StartsWith(UploadsDirectory, StringComparison.OrdinalIgnoreCase))
            {
                return BadRequest("Access denied.");
            }

            if (!System.IO.File.Exists(fullPath))
            {
                return NotFound("File not found.");
            }

            string content = System.IO.File.ReadAllText(fullPath);
            return Ok(content);
        }

        // GET /api/pathtraversalfixed/download?file=report.pdf
        [HttpGet("download")]
        public IActionResult DownloadFileBytes([FromQuery] string file)
        {
            // Apply the same validation as above
            if (string.IsNullOrWhiteSpace(file))
            {
                return BadRequest("Missing required parameter: file");
            }

            string sanitizedName = Path.GetFileName(file);

            if (file.Contains("..") ||
                file.Contains(Path.DirectorySeparatorChar) ||
                file.Contains(Path.AltDirectorySeparatorChar))
            {
                return BadRequest("Invalid filename.");
            }

            string fullPath = Path.GetFullPath(
                Path.Combine(UploadsDirectory, sanitizedName));

            if (!fullPath.StartsWith(UploadsDirectory, StringComparison.OrdinalIgnoreCase))
            {
                return BadRequest("Access denied.");
            }

            if (!System.IO.File.Exists(fullPath))
            {
                return NotFound("File not found.");
            }

            byte[] data = System.IO.File.ReadAllBytes(fullPath);
            return File(data, "application/octet-stream", sanitizedName);
        }
    }
}
